doc: Add Shell-Related documentation

ubieda opened the pull request:
### Description
Add Shell-Related Documentation:
- General
    - Add Backend configuration options for USB and Bluetooth LE.
    - Add Configuration for commonly-used Shell command groups (GPIO, I2C, Sensors, Flash, File-System).
- Shell module sample
    - Add first version of README describing different types of options registering Shell commands.

kartben commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,157 @@
+   :gen-args: -DCONFIG_SENSOR=y -DCONFIG_SENSOR_SHELL=y
+   :compact:
+
+Backend - USB CDC ACM
+=====================
+
+The application may be configured to expose Shell over USB CDC ACM.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/shell/shell_module
+   :host-os: unix
+   :board: nrf52840dk/nrf52840
+   :goals: flash
+   :gen-args: -DOVERLAY_CONFIG='overlay-usb.conf' -DDTC_OVERLAY_FILE='usb.overlay'
+   :compact:
+
+Backend - Shell over Bluetooth LE (NUS)
+=======================================
+
+The application may also be built for boards with Bluetooth capabilities, by
+exposing the shell interface over Bluetooth LE (NUS). The following commands
+are applicable to the nrf52840dk/nrf52840.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/shell/shell_module
+   :host-os: unix
+   :board: nrf52840dk/nrf52840
+   :goals: flash
+   :gen-args: -DOVERLAY_CONFIG='overlay-bt.conf' -DDTC_OVERLAY_FILE='bt.overlay'
+   :compact:
```
mmm shouldn't we just encourage people to use existing snippets and drop these overlays? They seem to be mostly a copy-paste of the equivalent snippets. I think these docs on "how to enable CDC/NUS/RTT shell" should be moved to the actual shell documentation page, as they have nothing to do with this sample specifically (even more so if the generic snippets can be used)

ubieda replied:
I've refactored the Shell documentation to reference the code-snippets (now shell-compatible for both `cdc-acm-console`, as well as for `nus-console`)

kartben commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,157 @@
+.. zephyr:code-sample:: shell-module
+   :name: Shell Sample
+   :relevant-api: shell_api
+
+   Basic sample for utilizing the Zephyr Shell
```
```suggestion
.. zephyr:code-sample:: shell-module
   :name: Custom shell module
   :relevant-api: shell_api

   Register shell commands using the Shell API
```

kartben reviewed (changes requested):
Thanks for adding this README - this will be one more to tick off of https://github.com/zephyrproject-rtos/zephyr/issues/27805 :)

However, this will likely need more work as the goal of the shell module sample is actually to demonstrate how to use the shell API for registering new commands, sub-commands, etc.

I like the idea of documenting somewhere the list of useful shell modules (GPIO, I2C, ...) and their associated KConfigs, but I don't think this belongs to this README.

Would you be able to rework this page so that it's indeed documenting the actual "custom shell module" aspects of it, and maybe move the instructions on how to enable CDC-ACM, NUS, etc. to the shell doc page?

Thanks!

jori-nordic reviewed (dismissed):
Thanks! I'm always looking for how to enable the various extra shell commands. 

I agree with @kartben that the various backends maybe should be in a more general shell page and linked in the README.
But not nacking, we can always move that later.

kartben commented:
> Thanks! I'm always looking for how to enable the various extra shell commands.

But again, this is *not* what this sample is meant for :) If we need (and we do, I think we all agree!) something like "Shell tips and tricks" then this belongs to the shell documentation, not the sample that's meant to explain how to actually write one's own shell commands.

ubieda commented:
@kartben @jori-nordic Thanks for your feedback - I agree with @kartben, I'll refactor this PR to capture those "tips and tricks" under the Shell documentation, and have the sample README expand on what the sample is doing.

ubieda commented:
@kartben sorry for taking a bit to iterate on this. Expect an update before the week ends!

ubieda committed:
doc: shell: Changes to restructure Backends documentation

- Unify shell backends in a dedicated section.
- Add documentation for the following backends: USB, BLE, RTT.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
doc: shell: Add section for commonly-used commands

With details on how to enable them.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: shell: shell_module: Add README

With basic description of registering different types of Shell commands.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
I've made a second pass of this PR, structuring the shell-related documentation in two sections:
- One on the existing Shell documentation, where details are captured on how to configure backends and commonly-used Shell commands.
- One on the Shell module sample, where some content regarding the registration of Shell commands is summarized.

However, as I was going through the documentation on the Shell sample, I realize the existing Shell documentation kind of captures most of the details on registering the different types of Shell commands. I did a first effort to capture a list on the README while avoiding content duplication, but there may be room for improvement, as well as desire to go a bit more in-depth in certain aspects.

I'm open to feedback as to what should go where and what could use more detail. As mentioned, this is another pass and could have more iterations.

jori-nordic commented on doc/services/shell/index.rst:
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
+
+To configure Shell USB CDC ACM backend:
+
+* Enable :kconfig:option:`CONFIG_SHELL_BACKEND_SERIAL`, along with other USB
```
I thought we had a snippet?

ubieda replied:
If I understand it correctly, the snippet cdc-acm-console is missing the zephyr,shell-uart redirection to the appropriate node.

I can file a separate PR to have that extra line included in the snippet.

LMk if I'm missing something!

jori-nordic replied:
ah alright, I was thinking of the console then. disregard the comment

ubieda replied:
Integrated now!

jori-nordic commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,200 @@
+==========================
+
+For more details on the Shell subsystem, check the general Shell documentation:
+:zephyr_file:`doc/services/shell/index.rst`
```
does linking to the rst section work from samples? I don't remember.
It's either sample -> general doc or general doc -> sample linking that doesn't.

ubieda replied:
I've changed it to **:ref:`Shell documentation <shell_api>`** and it does appear to work on a local build.

kartben commented on doc/services/shell/index.rst:
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
 `tab completion <tab-feature_>`_, and `history <history-feature_>`_
 features of the shell.
 
+USB CDC ACM Backend
+===================
```
it looks like backend "headings" are a bit all over the place (see USB CDC, Telnet, RTT, ...) can you have a look and see if at a minimum you can't group USB CDC and Telnet together under the same heading (and maybe move RTT there too, if appropriate?)

ubieda replied:
I've updated this branch attempting to gather all the already described backends. However, the list is not exhaustive and did not try to group using Shell as Logging backend: hopefully that's okay!

ubieda commented on doc/services/shell/index.rst:
```diff
@@ -112,6 +194,46 @@ types:
+.. code-block:: kconfig
+
+	CONFIG_FILE_SYSTEM=y
+	CONFIG_FILE_SYSTEM_SHELL=y
```
I've gotten rid of these code-blocks altogether. Instead, I've changed it to a list and made the options linkable to their Kconfig source.

ubieda commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,200 @@
+
+      /* Disable obscure mode */
+      shell_oscure_set(sh, false);
+      ...
```
Done!

ubieda commented on doc/services/shell/index.rst:
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
+	CONFIG_UART_LINE_CTRL=y
+
+* Assign the chosen property `zephyr,shell-uart` to a node compatible with
+  `zephyr,cdc-acm-uart`:
```
Done!

ubieda commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,200 @@
+      log-level  : <module name> <severity (0-4)>
+    uart:~$ kernel version
+    Zephyr version 3.6.99
+    uart:~$
```
I agree - Done!

ubieda commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,200 @@
+********
+
+This is a simple application demonstrating how to write and register commands
+using the Shell API.
```
Done!

ubieda commented on samples/subsys/shell/shell_module/README.rst:
```diff
@@ -0,0 +1,200 @@
+==========================
+
+For more details on the Shell subsystem, check the general Shell documentation:
+:zephyr_file:`doc/services/shell/index.rst`
```
Done!

ubieda commented on doc/services/shell/index.rst:
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
+
+.. code-block:: kconfig
+
+	# Enable Serial
```
Done - Hopefully I didn't miss anything

jakub-uC reviewed (dismissed):
Thank you

kartben commented on samples/subsys/shell/shell_module/README.rst:
````diff
@@ -0,0 +1,115 @@
+using the :ref:`Shell API <shell_api>`:
+
+Register Static commands
+   ```version``` is a static command that prints the kernel version.
````
double backquotes, not triple, please

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/b3bfa0c9-80bd-4968-8e27-04b5c1eae7b7)

ubieda replied:
Roger - got confused by this suggestion: https://github.com/zephyrproject-rtos/zephyr/pull/71904#discussion_r1603638630.

Addressed.

kartben replied:
> Roger - got confused by this suggestion: [#71904 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71904#discussion_r1603638630).

woops, sorry!

kartben commented on doc/services/shell/index.rst:
```diff
@@ -49,10 +54,76 @@ At this point, the following transport layers are implemented:
+USB CDC ACM
+===========
+
+To configure Shell USB CDC ACM backend, simply add the snippet `cdc-acm-console`
```
```suggestion
To configure Shell USB CDC ACM backend, simply add the snippet ``cdc-acm-console``
```

kartben commented on doc/services/shell/index.rst:
```diff
@@ -49,10 +54,76 @@ At this point, the following transport layers are implemented:
+Bluetooth LE (NUS)
+==================
+
+To configure Bluetooth LE (NUS) backend, simply add the snippet `nus-console`
```
```suggestion
To configure Bluetooth LE (NUS) backend, simply add the snippet ``nus-console``
```

kartben reviewed (changes requested):
this is brilliant, thanks @ubieda!
Some backticks fun and it should be good to go

simhein reviewed (approved):
Thanks, LGTM
